docs/design/2021-02-09-ttl-table.md

```diff
@@ -0,0 +1,45 @@
+
+## Proposal
+Introduce new `TTL` and `TTL_GRANULARITY` table options when creating or altering tables. By specifying `TTL`, users can expect expired data to be removed automatically. Additionally, users can also choose from `ROW` or `PARTITION` for `TTL_GRANULARITY` option to trade off collect granularity and cost to run such garbage collection. `ROW` mode evaluates expiry time against each row of data and reclaim space at row basis. This gives us the finest granularity and allows the most accurate expiration timing. `PARTITION` mode on the other hand partitions data according to its last update time. A background timer is going to rollover partitions and truncate all expired data within the oldest partition all at once.
+
+## Rationale
+Users can choose to store expiry time explicitly in the table schema and run delete on such tables periodically to reclaim space. This approach not only post burden on users but also introduced unnecessary load on TiDB to execute such updates and run GC at background to eventually reclaim space. The extra load is especially high when there is lots of data expiring within a short period of time.
+Another way to quickly remove expired data is managing all data in a carefully designed partition table. Bucketing data according to their update time makes it possible to quickly delete all expired records within a partition with a simple `TRUNCATE`. Compared to TTL Table's approach, these two options are error prone and suboptimal.
+
+## Compatibility and Migration Plan
+Due to the fact that TiDB by itself prohibits either converting ordinary table to partition table or between different partition table types, it is only possible to convert existing table to TTL table with `ROW` granularity. A TTL table with `PARTITION` granularity is implemented as a special partition table therefore conflicts with any other partition table type. Newly created tables can choose either `ROW` or `PARTITION` as a trade off between accuracy of reclaiming time and efficiency of garbage collecting expired data.
+
+## Implementation
+
+### `ROW`
+By associating TTL configuration with the key range of a TTL table in PD and distributing such configuration to all TiKV instances within the cluster, TiKV can utilize TTL settings during GC process to collect data that are expired. To avoid inconsistency caused by reclaiming progress difference between index and actual record data. The lifetime of record range is one GC interval longer than it is for index range, therefore TiDB will not see missing record data during table lookup.
```

Comment:
I'm not quite clear about the 'GC' here, for the expired data, is it possible to read them using the `tidb_snapshot`(https://docs.pingcap.com/tidb/stable/read-historical-data)?

Reply:
The `GC` process mentioned here is MVCC GC process. For this case, once the data are considered expiring and being reclaimed, users can not read them even with `tidb_snapshot` because data are physically deleted by `GC`.